base/src/main/java/io/spine/query/Distribution.java

```diff
@@ -0,0 +1,307 @@
+                                                        .build();
+                result.addExpression(childAnd);
+            }
+
+            for (CustomSubjectParameter<?, ?> secondCustom : second.customParams()) {
+                AndExpression.AndBuilder<R> childBuilder = AndExpression.newBuilder();
+                addChild(firstChild, childBuilder);
+                AndExpression<R> childAnd = childBuilder.addCustomParam(secondCustom)
+                                                        .build();
+                result.addExpression(childAnd);
+            }
+
+            for (Expression<R, ?> secondChild : second.children()) {
+                AndExpression.AndBuilder<R> childBuilder = AndExpression.newBuilder();
+                addChild(firstChild, childBuilder);
+                addChild(secondChild, childBuilder);
+                AndExpression<R> childAnd = childBuilder.build();
+                result.addExpression(childAnd);
+            }
+        }
+    }
+
+    @SuppressWarnings("MethodWithMultipleLoops")    /* Avoiding multiple similar methods. */
+    private static <R>
+    void cartesianCustomParams(List<CustomSubjectParameter<?, ?>> customOfFirst,
+                               OrExpression<R> second,
+                               OrExpression.OrBuilder<R> result) {
+        for (CustomSubjectParameter<?, ?> firstCustom : customOfFirst) {
+
+            for (SubjectParameter<R, ?, ?> secondParam : second.params()) {
+                AndExpression.AndBuilder<R> childBuilder = AndExpression.newBuilder();
+                AndExpression<R> childAnd = childBuilder.addCustomParam(firstCustom)
+                                                        .addParam(secondParam)
+                                                        .build();
+                result.addExpression(childAnd);
+            }
+
+            for (CustomSubjectParameter<?, ?> secondCustom : second.customParams()) {
+                AndExpression.AndBuilder<R> childBuilder = AndExpression.newBuilder();
+                AndExpression<R> childAnd = childBuilder.addCustomParam(firstCustom)
+                                                        .addCustomParam(secondCustom)
+                                                        .build();
+                result.addExpression(childAnd);
+            }
+
+            for (Expression<R, ?> secondChild : second.children()) {
+                AndExpression.AndBuilder<R> childBuilder = AndExpression.newBuilder();
+                addChild(secondChild, childBuilder);
+                AndExpression<R> childAnd = childBuilder.addCustomParam(firstCustom)
+                                                        .build();
+                result.addExpression(childAnd);
+            }
+        }
+    }
+
+    @SuppressWarnings("MethodWithMultipleLoops")    /* Avoiding multiple similar methods. */
+    private static <R>
+    void cartesianSimpleParams(List<SubjectParameter<R, ?, ?>> paramsOfFirst,
```

Comment:
This and previous method look similar. The methods itself are complex too. Please refactor. 

Consider having a helper class that does the Cartesian thing. As a side effect, you'd be able to extract nested `for` loops into methods.